### DISCUSSION TITLE
Best practice around data files

### DISCUSSION THREAD

**Discussion by juanmaia** [pyscaffold/pyscaffold#558] [Q&A] [answered]
Hey there, my question is related around setuptools, (data_files, package_dir) and how PyScaffold uses it.

When dealing with data files, at the root level of the repo, if in my code I read them with the relative path `../../data/file.csv/` after installing my module with `pip install .` python modules inside `src` will get copied to the root level of the package, thus making the way I read my data files break. 

ps: I would rather not move data files inside `src`

What's the best way around this?

**Comment by abravalheri** [comment, marked as answer]
Hi @juanmaia, unfortunately `data_files` and files outside the package directory are discouraged/deprecated in the package ecosystem...

There is no support for finding where they are installed, and the place they are copied will vary depending on the way the users will do it (e.g. installing with the root Python, `--user`, virtual env, conda). That is the main reason why they are not supported.

Also notice that files distributed with the package should be "read-only".

If you need writeable files, or files outside the package directory I recomend using `appdirs` to create them the first time your program runs.

You can have read-only template inside the package directory and then copy them to the location returned by `appdirs` using `shutil`.

(`appdirs` is a package in PyPi and `shutil` is part of the stdlib).

**Comment by FlorianWilhelm** [comment]
Hi @juanmaia, when building your package with the recommended `wheel` format, basically only what's in `src/YOUR_PACKAGE` plus some meta information will be packaged. This makes total sense since later on after installation you will reference this with `YOUR_PACKAGE` when importing and so on. Also your non-Python data files need to live somewhere in your packages structure since you need a way to reference them, e.g. with 
```
from importlib.resources import read_text, read_binary
data = read_text('YOUR_PACKAGE', 'file.csv')
```
You see that having your data files somewhere in your package structure is essential to actually work with them as you cannot rely on file system operations since `pip` and `virtualenv`/`conda` will do their magic and you don't know where the files end up.

**Comment by juanmaia** [comment]
Thank you both @FlorianWilhelm and @abravalheri, this was really helpful.

**Reply by FlorianWilhelm** [reply]
Thanks @juanmaia. Would you mind marking this question as answered?

**Reply by juanmaia** [reply]
@FlorianWilhelm I don't know how to do it. Can you please tell me?

**Reply by juanmaia** [reply]
After it was moved to "Q&A" the button to mark as Answered showed up!

**Comment by julie777** [comment]
@juanmaia 
I think my opinion differs from the others, because I have just spent a couple days doing nothing but reading PyPA docs.

They are pretty clear about packaging. 
- files that need to be distributed as part of a package should be in src. Src is the PyPI package/distribution source files (not source code files) for the project.
- setuptools provides for autodiscovery of items in src and can add them to the package without a MANIFEST file.
- if they are used at runtime, they are in a package relative directory.

Caveout, I haven't had to do a bunch of experimenting with data files, so you will have to read the [setuptools](https://setuptools.pypa.io/en/latest/) and give it a try.
